KAN-224-블로그

kimgorok opened the pull request:
## 😀 제목 : 블로그 페이지 api 연결

<br/>

## 🔎 작업 내용

- 어떤 변경 사항이 있었나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [x] CSS 등 사용자 UI 디자인 변경
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [ ] 코드 리팩토링
- [x] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [x] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

<br/>

## 🎨 이미지 첨부

<img width="1097" alt="스크린샷 2024-11-14 오후 4 04 47" src="https://github.com/user-attachments/assets/d2c7191a-0760-464f-b53d-c05625c3c884">

<br/>

## 🔧 앞으로의 과제

- 다음으로 할 일은 무엇인가요??
api 연결을 진행합니다.

  <br/>

## ➕ 추가 코멘트

### 🥨 무한스크롤

```typescript
  const { ref, inView } = useInView(); // 무한 스크롤을 위한 InterSection Observer 훅

   ...

  useCustomInfiniteQuery(['recommendPosts'], ({ pageParam = '0' }) => fetchPosts(pageParam), {
      getNextPageParam: (lastPage) => {
        if (lastPage.data.lastId === -1) return undefined; // lastId가 -1이면 마지막 페이지
        return lastPage.data.lastId; // 다음 배열의 lastId 반환
      },
      initialPageParam: '0', // 초기 페이지 파라미터
    }),
    
    ...
    
   // 무한 스크롤 처리
  useEffect(() => {
    // inView가 true이고 현재 쿼리의 다음 데이터가 있을 때
    if (inView && currentQuery) {
      const { hasNextPage, isFetchingNextPage, fetchNextPage } = currentQuery;
      if (hasNextPage && !isFetchingNextPage) {
        fetchNextPage();
      }
    }
  }, [inView, currentQuery]);
  
  ...
  
  {/* 무한 스크롤 트리거용 div */}
  <div ref={ref} className="h-1" />
```

무한스크롤은 다음과 같이 
```typescript
const { 감지영역, 화면에 들어왔는지감지함수 } = useInView();

...

useCustomInfiniteQuery([키값], (페이지파라미터) => 데이터를 불러오는 함수, {
  다음 페이지를 불러오는 옵션: ()  => { },
  초기 페이지,
}),

...

useEffect(()=>{
  if(사용자에 화면에 inView가 있는지) {
    다음 페이지 불러오는 함수
   }
}, [inView, 현재쿼리]);

...

<div ref={ref} />  // 감지 위치
```
이렇게 작성해주시면 됩니다.

<hr />

### 🥪 텍스트에디터

```typescript
// Next.js의 dynamic import를 사용하여 ToastEditor 컴포넌트를 동적으로 불러옴, 동적으로 불러올 컴포넌트의 경로를 지정
const ToastEditor = dynamic(() => import('../../../../../_components/ToastEditor'), {
  ssr: false, // 서버사이드 렌더링 비활성화
  loading: () => <div className="w-full h-96 bg-gray-200 rounded-md animate-pulse" />,
});

...

  const handleImage = useCallback(async (file: File, callback: (url: string) => void) => {
    try {
      const imageUrl = await postPic(file);
      callback(imageUrl);
    } catch (error) {
      console.error('이미지 업로드 실패:', error);
    }
  }, []);

...

  <ToastEditor
        initialValue={content}
        onChange={(value: string) => setContent(value)}
        height={editorHeight}
        handleImage={handleImage}
      />
```

텍스트 에디터는 다음과 같이
```typescript
const ToastEditor = dynamic(()=> import('토스트에디터 컴포넌트 경로'), {
   ssr: false, // 클라이언트사이드에서만 렌더링 되게
   loading: () => 로딩중에 보여줄 거
});

...

const 이미지처리함수 = useCallback(파일 타입, 리턴받을 url 타입) => {
   const url저장변수  = await 이미지업로드함수;
   callback(url저장변수); // url 콜백
}, []);

...

  <ToastEditor
     initailValue={토스트에디터 value state}
     onChange={토스트에디터 onChange함수}
     height={토스트에디터 높이값}
     handleImage={이미지 처리 함수}
/>
```
이렇게 작성해주시면 됩니다.

- [무한스크롤](https://velog.io/@kimgorok/Popolog-TanstackQuery-Intersection-Observer%EB%A5%BC-%EC%82%AC%EC%9A%A9%ED%95%B4-%EB%AC%B4%ED%95%9C-%EC%8A%A4%ED%81%AC%EB%A1%A4-%EA%B5%AC%ED%98%84%ED%95%98%EA%B8%B0)
- [텍스트에디터](https://velog.io/@kimgorok/Popolog-Toast-uireact-editor%EB%A5%BC-%EC%82%AC%EC%9A%A9%ED%95%B4-%EC%BB%A4%EC%8A%A4%ED%85%80-%ED%85%8D%EC%8A%A4%ED%8A%B8-%EC%97%90%EB%94%94%ED%84%B0-%EA%B5%AC%ED%98%84%ED%95%98%EA%B8%B0)


<hr />

- (11.14)
1. toast ui editor를 설치한 뒤 포스트 작성 페이지에 적용하였습니다.

- (11.15)
1. textEditor 로딩 중 스켈레톤 UI를 추가했습니다.

- (11.18)
1. textEditor에 이미지를 업로드하고 백엔드로 전송 후 이미지 url을 받는 함수를 추가했습니다.

- (11.19)
1. 사용자의 화면 높이에 따라 textEditor의 높이를 변화시키는 useWindowHeight 훅을 작성했습니다.

- (11.20)
1. useCustomMutation을 사용해 loveMutation, bookmarkMutation 함수를 추가했습니다.
2. onMutate와 onError를 추가하여 좋아요, 북마크 버튼에 낙관적 업데이트를 적용했습니다.
3. 포스트 수정, 삭제 버튼을 구현했습니다.

- (11.21)
1. url에 postId대신 postTitle이 들어가도록 수정했습니다.
2. URL로 포스트를 조회하는 함수를 추가했습니다.
3. 좋아요, 북마크를 표시하는 useLoveAndBookmark훅을 따로 분리했습니다.

- (11.22)
1. react-intersection-observer를 설치하고, useCustomInfininteQuery.ts를 작성하여 추천 피드에 무한 스크롤을 구현했습니다.

- (11.23)
1. 추천 탭 이외의 다른 탭에도 무한 스크롤을 적용했습니다.

- (11.25)
1. 댓글/답글 작성을 구현했습니다.

<br/>

## 🤔 Issue Ticket

- [Ticket](https://gwagseoyeon.atlassian.net/browse/KAN-224)

kimgorok committed:
🎉init: [블로그] ToastEditor 설치 및 적용

- ToastUiEditor 설치
- ToastEditor.tsx 생성
- 기존 글쓰기 페이지에 적용
- 기존 글쓰기 페이지의 react-hook-form 제거

KAN-224

kimgorok committed:
🐛fix: [블로그] ToastEditor 수정

- 동적 import 위치 수정
- 제목과 내용 입력 감지
- 타입 위치 변경
- change 이벤트코드 수정

KAN-224

kimgorok committed:
Merge branch 'KAN-224-블로그' of https://github.com/KEA-5th-Myaong/frontend into KAN-224-블로그

kimgorok committed:
Merge branch 'main' of https://github.com/KEA-5th-Myaong/frontend into KAN-224-블로그

kimgorok committed:
💄design [블로그] UI 수정

- 스켈레톤 UI 수정

KAN-224

kimgorok committed:
💄design: [블로그]  뒤로가기 버튼 위치 수정

- 뒤로가기 버튼의 px 제거

KAN-224

kimgorok committed:
💄design: [블로그] 필요없는 div 제거 및 UI 수정

- 필요없는 div 합침

KAN-224

kimgorok committed:
✨feat: [블로그] 텍스트에디터 이미지 업로드 함수 작성

- postPic 함수가 이미지 파일을 받도록 수정
- ToastEditor.tsx에 이미지 처리 함수 props 작성

KAN-224

yeonilil commented on src/app/_components/ToastEditor.tsx:
```diff
@@ -0,0 +1,41 @@
+  handleImage: (blob: File, callback: (url: string) => void) => Promise<void>;
+}
+
+export default function ToastEditor({ initialValue = '', onChange, height, handleImage }: ToastEditorProps) {
```
관리자 문의하기 페이지나 공지 작성 페이지에서도 이 에디터를 재사용 할 수 있을까요??

kimgorok replied:
넵 에디터를 사용하는 페이지에서 몇 가지 함수를 사용하면 재사용이 가능합니다.
그러나 에디터를 통해 작성된 데이터는, HTML로 변환되어 백엔드로 전송이 되기 때문에 백엔드분들과의 얘기가 필요할 거 같습니다.🤔

yeonilil replied:
백엔드 분들께 여쭤봤는데 문의 작성, 관리자 공지 작성 부분에서 사용해도 된다고 하셨습니다!

kimgorok committed:
🐛fix: [블로그] 이미지 전송 함수 수정

- formData를 전송하도록 수정

KAN-224

kimgorok committed:
✨feat: [블로그] 화면 높이에 따른 크기 조절 훅 생성 및 에디터에 적용

- useWindowHeight.tsx 훅을 생성하여 화면 높이 감지
- 화면 높이 감지 후 텍스트 에디터 크기 유동적으로 변경

KAN-224

kimgorok committed:
🐛fix: [블로그] 포스트 이동 범위 수정

- e.stopPropagation 적용으로 블로그 주인 주소 이동 적용

KAN-224

kimgorok committed:
✨feat: [블로그] 포스트 북마크, 좋아요 버튼 api 연결

- useCustomMutation훅을 사용한 loveMutation, bookmarkMutation 함수 추가

KAN-224

kimgorok committed:
✨feat: [블로그] 포스트 북마크, 좋아요 버튼 낙관적 업데이트 적용

- onMutate와 onError를 추가하여 낙관적 업데이트 반영

KAN-224

kimgorok committed:
✨feat: [블로그] 포스트 수정, 삭제 버튼 구현

- deletePost 연결 및 포스트 수정 함수 추가

KAN-224

kimgorok committed:
✨feat: [블로그] 포스트 함수 메모이제이션 추가

- useMemo와 useCallback 적용

KAN-224

kimgorok committed:
✨feat: [블로그] 포스트 작성/수정 분리

- 글쓰기 페이지에 '수정'버튼을 눌러서 진입했을 경우 isEdit을 true 한 뒤 조건부 렌더링 적용
- '수정'버튼을 눌러서 진입했는지 여부는 url 쿼리스트링을 통해 확인

KAN-224

kimgorok committed:
✨feat: [블로그] 글 제목도 store에 저장하도록 수정

- usePostWriteStore에 postTitle 속성 추가 및 적용

KAN-224

kimgorok committed:
🚚rename: [블로그] url을 postId에서 postTitle로 변경

- url에 postId를 표시하지 않도록 수정

KAN-224

kimgorok committed:
✨feat: [블로그] URL로 포스트 조회 추가

- 타인 정보 조회, URL로 포스트 조회 함수 추가

KAN-224

kimgorok committed:
♻️refactor: [블로그] 좋아요, 북마크 표시 훅 분리

- useLoveAndBookmark 훅 생성
- 재사용성 강화

KAN-224

kimgorok committed:
✨feat: [블로그] 포스트 상세에 loveMutation 훅 적용

- 실제 api 연결 후 추가 테스트 필요

KAN-224

kimgorok committed:
✨feat: [블로그] 추천 피드 무한스크롤 구현

- react-intersection-observer 설치
- 무한스크롤을 위한 훅 useCustomInfiniteQuery.ts 작성
- PostFeed 추천 피드에 작성

KAN-224

kimgorok committed:
🐛fix: [블로그] 피드 타입 오류 해결

- PostResponse 인터페이스 작성
- 각 데이터 호출 및 로딩 상태 간소화

KAN-224

kimgorok committed:
✨feat: [블로그] 메인페이지 다른 탭에도 무한 스크롤 적용

- 팔로잉, 북마크, 직군 탭에도 무한스크롤 적용
- 리팩토링 이전
- 타입 분리

KAN-224

kimgorok committed:
🐛fix: [블로그] 타입 수정

- PostDetailProps를 PostProps에 합침

KAN-224

kimgorok committed:
🐛fix: [블로그] useCustomInfiniteQuery.ts 타입 오류 해결

- 기본 값으로 initailPageParam 추가
- queryFn과 options에 추가 타입 지정

KAN-224

kimgorok committed:
♻️refactor: [블로그] 무한스크롤 구현 리팩토링

- 각 탭별 쿼리 설정 객체를 생성해 코드 단순화

KAN-224

kimgorok committed:
✨feat: [블로그] 메인 피드에 북마크 좋아요 버튼 훅 연결

- useLoveAndBookmark훅 연결

KAN-224

kimgorok committed:
✨feat: [블로그] 댓글 작성 함수 연결

- useCustomMutation을 활용한 낙관적 업데이트 적용

KAN-224

kimgorok committed:
✨feat: [블로그] 답글 작성 함수 연결

- postReplyMutation 작성
- 타입오류 해결 및 리팩토링 이전

KAN-224

kimgorok committed:
🐛fix: [블로그] 텍스트 에디터 클린업 함수 추가 및 이미지 함수 에러처리 추가

- instance를 off하는 클린업 함수 작성
- try-catch를 사용해 handleImage의 에러처리 추가

KAN-224

BaekJiyeon02 commented on src/app/(route)/(blog)/blog/[username]/[postTitle]/_components/PostContent.tsx:
```diff
@@ -0,0 +1,174 @@
+  const sanitizedContent = useMemo(
+    () => ({ __html: DOMPurify.sanitize(data?.data.content || '') }),
+    [data?.data.content],
+  );
```
DOMpurify를 쓰셨는데 과거 멘토링때 피드백 받은대로 내용에 html 태그가 포함되어 있을 시 무효화 시키는 기능인가요?

kimgorok replied:
네 맞습니다!

BaekJiyeon02 commented on src/app/(route)/(blog)/blog/[username]/[postTitle]/_components/_comment/PostComment.tsx:
```diff
@@ -0,0 +1,219 @@
+
+  const postReplyMutation = useCustomMutation(postReplies, {
+    onMutate: async (commentData: { postId: string; comment: string; parentId: string }) => {
+      // 진행 중인 쿼리를 취소
```
진행 중인 쿼리를 취소한다는게 혹시 어떤 말씀이신지 설명가능하실까요?🤔

kimgorok replied:
tanstack-query를 사용하면 데이터를 불러오는 쿼리를 key값을 통해 관리합니다. 그리고 설정한 쿼리의 key값을 취소하면 중복호출을 방지할 수 있습니다. 이 함수로 예를 들면 답글을 작성하는 `postReplyMutation` 함수가 실행중일 때 사용자가 이 함수를 한번 더 실행시킬 경우 이전의 함수를 취소하고 새로운 함수로 다시 실행시킬 수 있습니다.

만약 사용자의 네트워크가 느릴 경우, 답글이 다 작성되기 전 답글작성 버튼을 **와다다다다** 누를 경우에 함수가 중복 호출될 수 있는데, 같은 key를 가진 이전의 함수를 다 취소하고 마지막 함수만 실행되게 할 수 있습니다.

BaekJiyeon02 reviewed (approved):
고생하셨습니다! 👍
